deCONZ migrate setup fully to config entry

Kane610 opened the pull request:
## Description:

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

Kane610 committed:
Initial working config entry with discovery

balloob commented on homeassistant/components/deconz/__init__.py:
```diff
@@ -46,44 +38,37 @@
+        elif CONF_HOST in config[DOMAIN]:
+            deconz_config = config[DOMAIN]
+        if deconz_config and \
+           deconz_config[CONF_HOST] not in configured_hosts(hass):
```
Since you accept any hostname, not just IP, this check won't work because discovered bridges will have IP. For Hue as workaround I've restricted config to just IPs

Kane610 replied:
Right now I only care for 1 bridge. But I've prepared to identify bridge regardless of hostname or IP by always storing the bridgeid.

balloob commented on homeassistant/components/deconz/config_flow.py:
```diff
@@ -0,0 +1,127 @@
+                    title='deCONZ',
+                    data=self.deconz_config
+                )
+            else:
```
No need for else since you return in if

balloob commented on homeassistant/components/deconz/config_flow.py:
```diff
@@ -0,0 +1,127 @@
+        if CONF_API_KEY not in import_config:
+            self.deconz_config = import_config
+            return await self.async_step_link()
+        return self.async_create_entry(
```
You should verify that it not already exists since this is called every time a user configures via `configuration.yaml`

balloob commented on homeassistant/components/deconz/config_flow.py:
```diff
@@ -0,0 +1,127 @@
+        """Handle a deCONZ config flow start."""
+        from pydeconz.utils import async_discovery
+
+        if DOMAIN in self.hass.data:
```
Wrong check. Count the config entries for this domain.

Also, add this check to discovery and import.

Kane610 committed:
No need for else

Kane610 committed:
Make sure that imported config doesnt exist as a config entry

Kane610 committed:
Improve checks to make sure there is only instance of deconz

Kane610 committed:
Fix tests and add new tests

Kane610 committed:
Follow upstream changes
Fix case when discovery started ongoing config entry and user completes setup  from other path it was possible to complete discovered config entry as well

Kane610 committed:
Add test to make sure link doesn't bypass any check for only allowing one config entry

Kane610 committed:
Dont use len to determine an empty sequence

Kane610 committed:
Cleanup

Kane610 committed:
Allways get bridgeid to use as unique identifier for bridge

balloob commented on homeassistant/components/deconz/__init__.py:
```diff
@@ -46,44 +38,36 @@
+        _LOGGER.error(
+            "Config entry failed since one deCONZ instance already exists")
+        return False
+    result = await async_setup_deconz(hass, None, entry.data)
```
```python
return await async_setup_deconz(…)
```

balloob replied:
You might want to just pass the `entry` along if you want to use the forward entry logic to setup the platforms.

Kane610 replied:
Thats the next step to look into. But I'm trying to avoid too much extra fluff in each PR since you've previously advised me against it ;)

balloob commented:
Looks good!
